---
title: "Comparison of genes for different alignments of same read"
author: "Florian Bieberich"
date: "10/13/2017"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, message = FALSE)
require("knitr")
opts_knit$set(root.dir = "/Users/florian/Desktop/Cambridge_Bioinformatics/Pipeline_project/R_workspace/")
```

```{r a, echo=FALSE, include=FALSE}
source("https://bioconductor.org/biocLite.R")
biocLite("TxDb.Mmusculus.UCSC.mm10.knownGene")
biocLite("GenomicFeatures")
biocLite("AnnotationDbi")
library(TxDb.Mmusculus.UCSC.mm10.knownGene)
library(GenomicFeatures)
library(AnnotationDbi)
library(GenomicAlignments)
library(Rsamtools)
library(GenomicRanges)
```

Test if ONT sequences multiple transcripts in one read.

Alignment of fastq file to reference genome using Minimap2:
```{r y, eval=FALSE, engine='bash'}
./minimap2 -ax map-ont /nfs/research2/marioni/florian/RefSeq/Mus_musculus.GRCm38.dna.sm.toplevel.fa /nfs/research2/marioni/florian/chopped_c4_reads.fastq > align_chop_c4.sam
```

Read in SAM table that was previously transformed to .txt (in bash) and supplementary columns
```{r z, eval=FALSE, engine='bash'}
awk '{print $1,$2,$3,$4,$5,$6,$7,$8,$9,$10,$11}' alignment.sam > align_chop_c4.txt #first 11 columns in new .txt file
```

The first eleven columns are named in the typical SAM format. The 12th column contains the read length 
and the 13th the position for the end of the read.

```{r b}
align_chop_c4 = read.table("align_chop_c4.txt", skip=1, quote="", comment="", sep=" ", stringsAsFactors = FALSE)
colnames(align_chop_c4) <- c("QNAME", "FLAG", "RNAME", "POS", "MAPQ", "CIGAR", "RNEXT", "PNEXT", "TLEN", "SEQ", "QUAL")
Read_len <- cigarWidthAlongReferenceSpace(align_chop_c4[,6]) #extract length from CIGAR string without softclips
align_chop_c4[,3] <- gsub("^", "chr", align_chop_c4[,3])#extend RNAME with "chr" to fit GRanges format
align_chop_c4_len <- cbind(align_chop_c4, Read_len) #append to dataframe
End_read <- (as.integer(align_chop_c4_len[,4])+as.integer(align_chop_c4_len[,12]))
align_chop_c4_len <- cbind(align_chop_c4_len, End_read)
```


Preparation of GRanges object for overlap search. The start and end position of the reads are combined in an IRanges object, which in turn is combined with the chromosomal information of the reads in the final GRanges object.

```{r c}
chop_c4_pos <- cbind(as.integer(align_chop_c4_len[,4]),as.integer(align_chop_c4_len[,4])+as.integer(align_chop_c4_len[,12]) , as.integer(align_chop_c4_len[,12]))
chop_c4_pos[is.na(chop_c4_pos)] <- 0
IR_c4 <- IRanges(as.integer(chop_c4_pos[,1]), as.integer(chop_c4_pos[,2]))
GR_c4 <- GRanges(seqnames = align_chop_c4_len[,3], ranges = IR_c4)
names(GR_c4) <- align_chop_c4_len[,1]
```


Conversion of exon and gene data from annotation package into Biostring objects.

```{r d}
exon_data <- exons(TxDb.Mmusculus.UCSC.mm10.knownGene)
gene_id <- mapIds(TxDb.Mmusculus.UCSC.mm10.knownGene, keytype="EXONID", key= as.character(exon_data$exon_id), column="GENEID")
```


Overlap of reads from initial SAM file with the exons from the annotation package are made. findOverlaps compares start and end position and chromosomal information to find overlapping exons from the annotated gene package.

```{r e}
overlap_c4 <- findOverlaps(GR_c4, exon_data)
```


Gene ids are split by reads that overlap with related exons of the gene. 

```{r f}
by.read <- split(gene_id[subjectHits(overlap_c4)], names(GR_c4)[queryHits(overlap_c4)])
```


As preparation to find the intersection of gene ids for different alignments for the same read, gene id's are split by the respective overlapping alignments. 
Further, the grouped gene id's are splitted by the read name of the alignment groups of gene id's. 

```{r g}
by.align <- split(gene_id[subjectHits(overlap_c4)], queryHits(overlap_c4)) #splits gene_ids by queryHits (alignments of reads)
by.read_align <- split(by.align, names(GR_c4)[as.integer(names(by.align))]) #names queryHits after matching read name
by.ra_filt <- by.read_align[lengths(by.read_align)>1] #filters reads with more than one alignment
```


Read names that have no intersection for the gene overlaps of their alignments could be the result of a false transcript sequencing.

```{r h}
by.intersec <- lapply(by.ra_filt, FUN=Reduce, f=intersect) #common gene_ids in different alignemnts for the same read
head(names(by.intersec[lengths(by.intersec)<1]), 10) #filters reads without common gene_id between the different alignments 
```


To investigate this assumption of different gene ids for different alignments from the same read, different gene names of the respective read have to be compared by using a database like the UCSC Genome Browser.

```{r i}
by.read_align$SRR5286963.39583 #alignments with different gene ids
align_chop_c4_len[align_chop_c4_len$QNAME=="SRR5286963.39583",c(1,2,3,4,5,6)] #SAM information about the alignments e.g. CIGAR
```















