---
title: "Pipeline - Read-gene mapping, adaptor chopping and clustering of overlapping reads"
author: "Florian Bieberich"
date: "10/13/2017"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


This pipeline is suited for ONT users to enable RNAseq downstream analysis. 
Reads are quality checked by first testing their sequencing quality (determine overlap of read with one or more genes), further chopping off UMI adaptor sequences and retrieving the UMI sequence and finally grouping them into cluster of reads that overlap each other and therefore should comprise (different) transcripts from the same gene.

The final application of this pipeline should read in FASTQ files and give out groups of reads that comprise the same transcipt (isoform) of one gene respectively ordered for each cell and should produce a consensus sequence for each group.


# Preparation

Linux, Windows or macOS, R and bash.


# Comparison of alignments from same read with overlapping gene(s)

This code provides a test for read quality from ONT sequencing data. 
Various alignments of the same read result of the hard- and softclipped regions. 
These are tested for their respective gene. 
Alignment of one read to multiple genes can implicate a failure during the sequencing process e.g. continuous sequencing of two different transcripts without pause in between. 

## How to 
Preparation: SAM file needs to be formatted to .txt.

Columns with length and end position of the reads are addded. 
GRanges object is constructed with RNAME column for seqnames and an IRanges object for the ranges, which in turn is made with start and end positions of the reads. 
The gene annotation package is prepared for further analysis bz extracting gene_id and exon_data. 
Now, overlaps between the reads and the exons of the gene annotation packaga are determined. Gene id's were splitted by the reads that overlap with a related exon. 
To further analyse the overlap of various alignments from the same read, gene id's were also splitted by the alignments that overlapped with a related exon. 
The resulting list was additionally splitted by the matching read name for each alignment (list of vectors, whereby vectors are the read names with the alignment - gene id touple as element). 
Finally, the intersect of the gene id's from the alignments of each read was formed and the resulting list was filtered for less than one intersect - indicates alignments with no common gene id!


# Adaptor chopping

Starting with a FASTQ file, adaptors with given sequence are identified and chopped off. 
The output is a FASTQ file containing the embedded part of the read (without adaptor).

## How to

Local-global pairwise alignment of adaptor sequence and the reads of the FASTQ file. 
Start and end position and length of adaptor aligned to each read is retrieved and used to identify embedded read (adaptor has to align with more than 95%). 
Sequence and quality score of embedded reads are formatted, originally named and finally written in a FASTQ file. 


# Clustering of overlapping reads

Aligned reads are overlapped to cluster them into subgroups.

## How to
Preparation: First six columns of SAM file need to be formatted to .txt.

Columns with read and query length, excluding and including softclipped region respectively are added. 
GRanges object was constructed with position, length, name and strand information. 
Reads that overlap with more than 99 bases with another read were detected. 
The reads that are overlapped by the same read were split. 
Subsequently, reads that overlap with the same read were marked with the same id and if a read that was already marked with another id was also included in the overlap, all reads with the same id as this read were connected to the new id by changing their cluster ids to the "current" one.















